### ISSUE TITLE
BUG: merge outer join drops rows with NaN datetime in index

### ISSUE THREAD

**Issue by realdanurbano** [pandas-dev/pandas#42536]
- [Yes] I have checked that this issue has not already been reported.

- [Yes] I have confirmed this bug exists on the latest version of pandas.
Tested with pandas 1.3.0. Verified the expected output is correct in the previous version, 1.2.5.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
import pandas as pd

left = ({'name': 'Foo', 'datetime': None, 'some column': 'some value'},)
left = pd.DataFrame(left).set_index(['name', 'datetime'])

right = ({'name': 'Foo', 'datetime': '2021-07-14', 'other column': 'other value'},)
right = pd.DataFrame(right).set_index(['name', 'datetime'])

merged = left.merge(right, how='outer', left_index=True, right_index=True)

print(merged)
```

#### Problem description
When merging left and right together using an OUTER join, I would expect to see two rows in the result. The problem appears to be with the `datetime` index column with `NaN` values. If I provide a date in left, I see the expected output. Without it, any row from the right is dropped, i.e. behaving like a left join.

Actual Output
```
              some column other column
name datetime                         
Foo  NaN       some value          NaN
```

#### Expected Output
```
                             some column other column
name datetime                           
Foo  NaN             some value          NaN
     2021-07-14    NaN  other value
```

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : f00ed8f47020034e752baf0250483053340971b0
python           : 3.8.1.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.6.0
Version          : Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.3.0
numpy            : 1.21.0
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 21.1.3
setuptools       : 41.2.0
Cython           : None
pytest           : 6.2.3
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.2
html5lib         : None
pymysql          : 1.0.2
psycopg2         : 2.9.1 (dt dec pq3 ext lo64)
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.1
numexpr          : None
odfpy            : None
openpyxl         : 3.0.7
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.2
sqlalchemy       : 1.4.4
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : None

</details>

**Comment by simonjayhawkins** [comment]
> * Verified the expected output is correct in the previous version, 1.2.5.

@realdanurbano on 1.2.5 (and earlier) I'm not seeing expected output, although the datetime value does appear in the output, so there has been a change, but since the output from previous versions was incorrect I'm not sure if this should be considered a regression.

```
>>> merged = left.merge(right, how="outer", left_index=True, right_index=True)
/home/simon/miniconda3/envs/pandas-1.2.5/lib/python3.9/site-packages/pandas/core/indexes/multi.py:3587: RuntimeWarning: The values in the array are unorderable. Pass `sort=False` to suppress this warning.
  uniq_tuples = lib.fast_unique_multiple([self._values, other._values], sort=sort)
>>> 
>>> print(merged)
                some column other column
name datetime                           
Foo  NaN         some value          NaN
     2021-07-14  some value  other value
>>> 
```

**Comment by realdanurbano** [comment]
I am not sure if I am missing something here, or perhaps my explanation is off or confusing, but it looks like you have confirmed that the expected output works in pandas `1.2.5`, which I agree with.  If you try this same example in versions `1.3.0` or `1.3.1`, you will not see the expected output.

**pandas version 1.2.5 shows the expected output**
```
In [4]: left.merge(right, how='outer', left_index=True, right_index=True)
/Users/DUrbano/.pyenv/versions/pandas-1-2-5/lib/python3.9/site-packages/pandas/core/indexes/multi.py:3587: RuntimeWarning: The values in the array are unorderable. Pass `sort=False` to suppress this warning.
  uniq_tuples = lib.fast_unique_multiple([self._values, other._values], sort=sort)
Out[4]: 
                some column other column
name datetime                           
Foo  NaN         some value          NaN
     2021-07-14  some value  other value
```
**pandas versions 1.3.0 and 1.3.1 shows the error output - i.e. the dropped row**
```
# pandas version 1.3.0
In [4]: left.merge(right, how='outer', left_index=True, right_index=True)
Out[4]: 
              some column other column
name datetime                         
Foo  NaN       some value          NaN
```
```
# pandas version 1.3.1
In [4]: left.merge(right, how='outer', left_index=True, right_index=True)
Out[4]: 
              some column other column
name datetime                         
Foo  NaN       some value          NaN
```

**Comment by realdanurbano** [comment]
I updated my last comment to help clear up some confusion that I introduced. Hopefully this is a bit clearer. ;)

**Comment by realdanurbano** [comment]
This issue is fixed as of pandas version 1.4.0. I also confirmed it fixed with the latest version, 1.4.2.

Thank you!
